protocol/immutable-flattening.md

```diff
@@ -0,0 +1,25 @@
+# Purpose
+
+The current fault proof system requires a redeployment of the `DisputeGame.sol`, `DisputeGameFactory.sol`, and `AnchorStateRegistry.sol`. Due to the immutable arguments in the constructor, specifically `_weth`, `_anchorStateRegistry`, and `_l2ChainId` the implementations must be fully redeployed for each new chain. The aim of this proposal is to simplify the Fauly Dispute Game contract setup so that all new deployments simply just require new proxies rather than new deployments of the factory and implementation.
+
+# Summary
+
+The `DisputeGameFactory.sol` should be upgraded to contain the arguments needed for every new `DisputeGame.sol`, and all of the constructor argument/logic should be moved into proxy which is actually pointed toward the `DisputeGame.sol` implementation, so that a single implementation of `DisputeGame.sol` can be used across each rollup.
+
+# Problem Statement + Context
+
+The Fault Dispute Game contracts are currently made up of a couple core contracts, `AnchorStateRegistry.sol`, `DelayedWETH.sol`, `DisputeGameFactory.sol`, and `FaultDisputeGame.sol/PermissionedFaultDisputeGame.sol`. When deploying "standard" rollups we currently need to deploy a new set of all of these contracts because specific immutable arguments in the implementation of `FaultDisputeGame.sol`, which in turn requires a specific `FaultDisputeGameFactory.sol` and in turn `AnchorStateRegistry.sol`. If `FaultDisputeGame.sol` implementations did not have chain specific immutables, then they could simply be added at the time of cloning on a per chain basis, requiring only one set of deployments and making the process simpler.
+
+An important nuance here is that each `FaultDisputeGame.sol` deployment has both a minimal proxy with immutable args deployed that points at an implementation specific to the `FaultDisputeGameFactory.sol`. The two sets of immutable args are first specific to the fault dipuste game, then the second set on the implementation are specific to the rollup.
+ 
+# Proposed Solution
+
+In order to elegantly resolve this, the two sets of immutable args should be flattened into a single set of immutable args deployed alongside the proxy from the factory. The factory will instead store all of the different per-chain values, and pass in all arguments as a top level proxy with immutable args.
```

Comment:
The factory can pass through the L2 chain ID but I don't think it's appropriate for the DisputeGameFactory to know of details of specific fault proof implementations like the absolute prestate. Not all game types will need that and DisputeGameFactory should be generic enough to work with any game implementation.

Reply:
It is possible to store all constructor args as just generic bytes, which while kind of messy because it hides the actual constructor args (we just store Implementations and byte arrays to pass in as "base" constructor args) I think its the only way to have generic chain-indifferent implementation contracts to clone from each type, but still be adaptable to multiple L2 deployments. 

Is there another approach you had in mind for flattened immutable args on games?

Reply:
> Not all game types will need that and DisputeGameFactory should be generic enough to work with any game implementation

What if the constructor args were stored within the DGF, in a flexible way. So, the current `gameImpls` mapping just holds the addresses of the impls:

```solidity
mapping(GameType => IDisputeGame) public gameImpls;
```

but could be updated to hold the constructor args too:

```solidity
struct GameData {
	IDisputeGame gameImpl;
	
}
mapping(GameType => GameData) public gameImpls;
```

Note that this breaks the `gameImpls()` interface, but gives an idea of one approach that avoids baking in game specific details.

Reply:
I'm not a huge fan of "here's a random binary blob that configures critical aspects of the dispute game", but this is logically very close to the approach that we're planning where the `DisputeGameFactory` holds a list of contracts that know how to create specific dispute game types (https://github.com/ethereum-optimism/design-docs/pull/153#issuecomment-2487065731). So rather than creating directly it calls those contracts to create the new game contract.  Then they can hold whatever game specific configuration they need in an understandable format.  Things like absolutePrestate can easily be written to storage instead of being an immutable and so we can follow the MCP pattern and hard forks could be done by just calling `setAbsolutePrestate` and not need to deploy any new contracts.

Reply:
So, the `DisputeGameFactory` will make reference to a set of ~game~ proposal type specific factories? 

I find that somewhat confusing, but I can get over it if it means that the proposal factories are MCP compatible, and we don't actually need to worry about the proposals being proxied to some single implementation (right?).

Reply:
OK, I found the [comment here](https://github.com/ethereum-optimism/design-docs/pull/153#issuecomment-2487065731). Noting that this means that N game types will mean N implementations and N proxies per system.

Reply:
> So, the `DisputeGameFactory` will make reference to a set of ~game~ proposal type specific factories?
> 
> I find that somewhat confusing, but I can get over it if it means that the proposal factories are MCP compatible, and we don't actually need to worry about the proposals being proxied to some single implementation (right?).

Do you think this is too much for OPCM? It's a lot of contracts/implementations/deployments but I think it ultimately makes it easier to reason about a lot of chains and should simplify our upgrade process from here a lot

Reply:
I've been going back and forth on this but I do ultimately think that the Creator pattern is the most flexible while also being the safest. It's much easier to verify things about a contract than it is to verify things about a data blob.

Reply:
The more I think about it the more I realize that the Creator pattern is roughly equivalent in complexity to the `gameImpl` CWIA pattern. Each game has either a gameImpl or a Creator that the DGF uses for game creation.